File: packages/sdk/clawdstrike-hunt/src/export.ts
/**
 * Export adapters — webhook, Splunk HEC, Elasticsearch, STIX 2.1, and CSV.
 */

import { ExportError } from './errors.js';
import type { Alert, TimelineEvent, IocMatch, IocEntry } from './types.js';

export interface RetryConfig {
  maxRetries?: number;
  baseDelayMs?: number;
}

export interface ExportAdapter {
  export(items: (Alert | TimelineEvent)[]): Promise<void>;
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return typeof value === 'object' && value !== null
    ? (value as Record<string, unknown>)
    : undefined;
}

function summarizeElasticBulkFailure(payload: unknown): string | undefined {
  const body = asRecord(payload);
  if (!body || body.errors !== true) {
    return undefined;
  }

  const items = Array.isArray(body.items) ? body.items : [];
  let total = 0;
  let failed = 0;
  let firstFailure: string | undefined;

  for (const item of items) {
    const itemRecord = asRecord(item);
    if (!itemRecord) {
      continue;
    }

    for (const [action, result] of Object.entries(itemRecord)) {
      const actionResult = asRecord(result);
      total += 1;
      if (!actionResult) {
        continue;
      }

      const status = typeof actionResult.status === 'number'
        ? actionResult.status
        : undefined;
      const itemError = actionResult.error;
      const hasError = itemError !== undefined || (status !== undefined && status >= 300);

      if (!hasError) {
        continue;
      }

      failed += 1;
      if (firstFailure === undefined) {
        const errorText = stringifyElasticError(itemError);
        const statusText = status !== undefined ? `status=${status}` : 'status=unknown';
        firstFailure = `${action} ${statusText}${errorText ? `: ${errorText}` : ''}`;
      }
    }
  }

  if (failed > 0) {
    return `${failed}/${total || failed} bulk items failed${firstFailure ? ` (${firstFailure})` : ''}`;
  }

  return 'bulk response reported errors=true';
}

function stringifyElasticError(value: unknown): string | undefined {
  if (typeof value === 'string') {
    return value;
  }

  const obj = asRecord(value);
  if (!obj) {
    return undefined;
  }

  const type = typeof obj.type === 'string' ? obj.type : undefined;
  const reason = typeof obj.reason === 'string' ? obj.reason : undefined;
  if (type && reason) {
    return `${type}: ${reason}`;
  }
  return type ?? reason;
}

async function withRetry(
  fn: () => Promise<Response>,
  config: RetryConfig | undefined,
): Promise<Response> {
  const maxRetries = config?.maxRetries ?? 0;
  const baseDelayMs = config?.baseDelayMs ?? 1000;

  let lastError: unknown;
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      const response = await fn();
      if (response.ok || response.status < 500) {
        return response;
      }
      // 5xx — retryable
      lastError = new ExportError(
        `Server error: ${response.status} ${response.statusText}`,
      );
    } catch (err) {
      // Network error — retryable
      lastError = err;
    }

    if (attempt < maxRetries) {
      await new Promise((resolve) =>
        setTimeout(resolve, baseDelayMs * 2 ** attempt),
      );
    }
  }
  if (lastError instanceof ExportError) throw lastError;
  throw new ExportError(
    `Export failed after ${maxRetries + 1} attempts: ${lastError instanceof Error ? lastError.message : String(lastError)}`,
  );
}

export class WebhookAdapter implements ExportAdapter {
  private url: string;
  private headers: Record<string, string>;
  private retry?: RetryConfig;

  constructor(url: string, headers?: Record<string, string>, retry?: RetryConfig) {
    this.url = url;
    this.headers = headers ?? {};
    this.retry = retry;
  }

  async export(items: (Alert | TimelineEvent)[]): Promise<void> {
    const body = JSON.stringify(items.map(itemToJSON));
    const response = await withRetry(
      () =>
        fetch(this.url, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json', ...this.headers },
          body,
        }),
      this.retry,
    );
    if (!response.ok) {
      throw new ExportError(
        `Webhook export failed: ${response.status} ${response.statusText}`,
      );
    }
  }
}

export class SplunkHECAdapter implements ExportAdapter {
  private url: string;
  private token: string;
  private index?: string;
  private retry?: RetryConfig;

  constructor(url: string, token: string, index?: string, retry?: RetryConfig) {
    this.url = url;
    this.token = token;
    this.index = index;
    this.retry = retry;
  }

  async export(items: (Alert | TimelineEvent)[]): Promise<void> {
    const events = items.map((item) => {
      const data = itemToJSON(item);
      const event: Record<string, unknown> = { event: data };
      if (this.index) event.index = this.index;
      return JSON.stringify(event);
    });
    const body = events.join('\n');

    const response = await withRetry(
      () =>
        fetch(this.url, {
          method: 'POST',
          headers: {
            Authorization: `Splunk ${this.token}`,
            'Content-Type': 'application/json',
          },
          body,
        }),
      this.retry,
    );
    if (!response.ok) {
      throw new ExportError(
        `Splunk HEC export failed: ${response.status} ${response.statusText}`,
      );
    }
  }
}

export class ElasticAdapter implements ExportAdapter {
  private url: string;
  private index: string;
  private apiKey?: string;
  private retry?: RetryConfig;

  constructor(url: string, index: string, apiKey?: string, retry?: RetryConfig) {
    this.url = url;
    this.index = index;
    this.apiKey = apiKey;
    this.retry = retry;
  }

  async export(items: (Alert | TimelineEvent)[]): Promise<void> {
    const lines: string[] = [];
    for (const item of items) {
      lines.push(JSON.stringify({ index: { _index: this.index } }));
      lines.push(JSON.stringify(itemToJSON(item)));
    }
    const body = lines.join('\n') + '\n';

    const headers: Record<string, string> = {
      'Content-Type': 'application/x-ndjson',
    };
    if (this.apiKey) {
      headers['Authorization'] = `ApiKey ${this.apiKey}`;
    }

    const response = await withRetry(
      () =>
        fetch(`${this.url}/_bulk`, {
          method: 'POST',
          headers,
          body,
        }),
      this.retry,
    );
    if (!response.ok) {
      throw new ExportError(
        `Elasticsearch export failed: ${response.status} ${response.statusText}`,
      );
    }

    let payload: unknown;
    try {
      payload = await response.json();
    } catch {
      throw new ExportError('Elasticsearch export failed: invalid _bulk JSON response');
    }

    const failureSummary = summarizeElasticBulkFailure(payload);
    if (failureSummary !== undefined) {
      throw new ExportError(`Elasticsearch export failed: ${failureSummary}`);
    }
  }
}

/**
 * Convert alerts and optional IOC matches to a STIX 2.1 bundle.
 */
export function toStix(
  alerts: Alert[],
  iocMatches?: IocMatch[],
): Record<string, unknown> {
  const objects: Record<string, unknown>[] = [];

  for (const alert of alerts) {
    objects.push({
      type: 'indicator',
      spec_version: '2.1',
      id: `indicator--${generateId()}`,
      created: alert.triggeredAt.toISOString(),
      modified: alert.triggeredAt.toISOString(),
      name: alert.title,
      description: alert.description,
      pattern_type: 'clawdstrike',
      pattern: `[alert:rule_name = '${escapeStixValue(alert.ruleName)}']`,
      valid_from: alert.triggeredAt.toISOString(),
      labels: [alert.severity],
    });
  }

  if (iocMatches) {
    for (const match of iocMatches) {
      for (const ioc of match.matchedIocs) {
        objects.push({
          type: 'indicator',
          spec_version: '2.1',
          id: `indicator--${generateId()}`,
          created: match.event.timestamp.toISOString(),
          modified: match.event.timestamp.toISOString(),
          name: `IOC: ${ioc.indicator}`,
          description: ioc.description ?? `IOC match: ${ioc.indicator}`,
          pattern_type: 'stix',
          pattern: iocToStixPattern(ioc),
          valid_from: match.event.timestamp.toISOString(),
        });
      }
    }
  }

  return {
    type: 'bundle',
    id: `bundle--${generateId()}`,
    objects,
  };
}

/**
 * Convert items to CSV string. Handles alerts, events, and mixed arrays.
 */
export function toCSV(items: (Alert | TimelineEvent)[]): string {
  if (items.length === 0) return '';

  const alerts = items.filter(isAlert);
  const events = items.filter((item): item is TimelineEvent => !isAlert(item));

  const sections: string[] = [];

  if (alerts.length > 0) {
    const headers = [
      'ruleName',
      'severity',
      'title',
      'triggeredAt',
      'description',
      'evidenceCount',
    ];
    const rows = alerts.map((a) =>
      [
        a.ruleName,
        a.severity,
        a.title,
        a.triggeredAt.toISOString(),
        a.description,
        String(a.evidence.length),
      ]
        .map(csvEscape)
        .join(','),
    );
    sections.push([headers.join(','), ...rows].join('\n'));
  }

  if (events.length > 0) {
    const headers = [
      'timestamp',
      'source',
      'kind',
      'verdict',
      'summary',
      'process',
      'actionType',
    ];
    const rows = events.map((e) =>
      [
        e.timestamp.toISOString(),
        e.source,
        e.kind,
        e.verdict,
        e.summary,
        e.process ?? '',
        e.actionType ?? '',
      ]
        .map(csvEscape)
        .join(','),
    );
    sections.push([headers.join(','), ...rows].join('\n'));
  }

  return sections.join('\n');
}

/**
 * Convert items to JSON Lines (JSONL) format — one JSON object per line.
 */
export function toJSONL(items: (Alert | TimelineEvent)[]): string {
  return items.map((item) => JSON.stringify(itemToJSON(item))).join('\n');
}

function isAlert(item: unknown): item is Alert {
  return (
    typeof item === 'object' &&
    item !== null &&
    'ruleName' in item &&
    'triggeredAt' in item
  );
}

function itemToJSON(item: Alert | TimelineEvent): Record<string, unknown> {
  if (isAlert(item)) {
    return {
      type: 'alert',
      ruleName: item.ruleName,
      severity: item.severity,
      title: item.title,
      triggeredAt: item.triggeredAt.toISOString(),
      description: item.description,
      evidenceCount: item.evidence.length,
    };
  }
  return {
    type: 'event',
    timestamp: item.timestamp.toISOString(),
    source: item.source,
    kind: item.kind,
    verdict: item.verdict,
    summary: item.summary,
    process: item.process,
    actionType: item.actionType,
  };
}

function csvEscape(value: string): string {
  if (value.includes(',') || value.includes('"') || value.includes('\n') || value.includes('\r')) {
    return '"' + value.replace(/"/g, '""') + '"';
  }
  return value;
}

function escapeStixValue(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/'/g, "\\'").replace(/]/g, '\\]');
}

function iocToStixPattern(ioc: IocEntry): string {
  const v = escapeStixValue(ioc.indicator);
  switch (ioc.iocType) {
    case 'sha256':
      return `[file:hashes.'SHA-256' = '${v}']`;
    case 'sha1':
      return `[file:hashes.'SHA-1' = '${v}']`;
    case 'md5':
      return `[file:hashes.MD5 = '${v}']`;
    case 'domain':
      return `[domain-name:value = '${v}']`;
    case 'ipv4':
      return `[ipv4-addr:value = '${v}']`;
    case 'ipv6':
      return `[ipv6-addr:value = '${v}']`;
    case 'url':
      return `[url:value = '${v}']`;
    default:
      return `[x-clawdstrike:value = '${v}']`;
  }
}

function generateId(): string {
  // Use crypto.randomUUID() when available, fall back to random hex.
  if (typeof globalThis.crypto?.randomUUID === 'function') {
    return globalThis.crypto.randomUUID();
  }
  const bytes = new Uint8Array(16);
  globalThis.crypto.getRandomValues(bytes);
  const hex = Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

File: packages/sdk/hush-py/src/clawdstrike/hunt/export.py
"""Export adapters for Hunt SDK -- webhook, Splunk HEC, Elasticsearch, STIX, CSV."""

from __future__ import annotations

import asyncio
import json
import uuid
from dataclasses import dataclass
from typing import Any, Callable, Coroutine, Protocol, Union, runtime_checkable

from clawdstrike.hunt.errors import ExportError
from clawdstrike.hunt.types import Alert, IocEntry, IocMatch, TimelineEvent


@dataclass
class RetryConfig:
    """Configuration for retry behaviour on export adapters."""

    max_retries: int = 0
    base_delay: float = 1.0


async def _with_retry(
    fn: Callable[..., Coroutine[Any, Any, Any]],
    retry: RetryConfig,
) -> Any:
    """Execute *fn* with exponential back-off on 5xx / connection errors.

    4xx errors are **not** retried (they indicate a client-side problem).
    """
    last_exc: Exception | None = None
    for attempt in range(1 + retry.max_retries):
        try:
            return await fn()
        except ExportError as exc:
            msg = str(exc)
            # Only retry on 5xx status codes
            status = _extract_status(msg)
            if status is not None and status < 500:
                raise
            last_exc = exc
        except Exception as exc:
            # Connection / transport errors are retryable
            last_exc = exc
        if attempt < retry.max_retries:
            await asyncio.sleep(retry.base_delay * (2 ** attempt))
    if last_exc is not None:
        raise last_exc
    raise ExportError("retry exhausted with no result")  # pragma: no cover


def _extract_status(msg: str) -> int | None:
    """Best-effort extraction of HTTP status code from an ExportError message."""
    # Messages look like "Webhook export failed: 503"
    parts = msg.rsplit(":", 1)
    if len(parts) == 2:
        try:
            return int(parts[1].strip())
        except ValueError:
            pass
    return None


@runtime_checkable
class ExportAdapter(Protocol):
    """Protocol for export adapters."""

    async def export(self, items: list[Union[Alert, TimelineEvent]]) -> None:
        pass


class WebhookAdapter:
    """Export items via HTTP POST to a webhook URL."""

    def __init__(
        self,
        url: str,
        headers: dict[str, str] | None = None,
        retry: RetryConfig | None = None,
    ) -> None:
        self.url = url
        self.headers = headers or {}
        self.retry = retry or RetryConfig()

    async def export(self, items: list[Union[Alert, TimelineEvent]]) -> None:
        try:
            import httpx
        except ImportError:
            raise ImportError(
                "httpx is required for WebhookAdapter: pip install httpx"
            )

        async def _do() -> None:
            body = json.dumps([_item_to_json(item) for item in items])
            async with httpx.AsyncClient() as client:
                resp = await client.post(
                    self.url,
                    content=body,
                    headers={"Content-Type": "application/json", **self.headers},
                )
                if resp.status_code >= 400:
                    raise ExportError(f"Webhook export failed: {resp.status_code}")

        await _with_retry(_do, self.retry)


class SplunkHECAdapter:
    """Export items to Splunk via the HTTP Event Collector."""

    def __init__(
        self,
        url: str,
        token: str,
        index: str | None = None,
        retry: RetryConfig | None = None,
    ) -> None:
        self.url = url
        self.token = token
        self.index = index
        self.retry = retry or RetryConfig()

    async def export(self, items: list[Union[Alert, TimelineEvent]]) -> None:
        try:
            import httpx
        except ImportError:
            raise ImportError(
                "httpx is required for SplunkHECAdapter: pip install httpx"
            )

        async def _do() -> None:
            events: list[str] = []
            for item in items:
                data = _item_to_json(item)
                event: dict = {"event": data}
                if self.index:
                    event["index"] = self.index
                events.append(json.dumps(event))
            body = "\n".join(events)
            async with httpx.AsyncClient() as client:
                resp = await client.post(
                    self.url,
                    content=body,
                    headers={
                        "Authorization": f"Splunk {self.token}",
                        "Content-Type": "application/json",
                    },
                )
                if resp.status_code >= 400:
                    raise ExportError(f"Splunk HEC export failed: {resp.status_code}")

        await _with_retry(_do, self.retry)


class ElasticAdapter:
    """Export items to Elasticsearch via the _bulk API."""

    def __init__(
        self,
        url: str,
        index: str,
        api_key: str | None = None,
        retry: RetryConfig | None = None,
    ) -> None:
        self.url = url
        self.index = index
        self.api_key = api_key
        self.retry = retry or RetryConfig()

    async def export(self, items: list[Union[Alert, TimelineEvent]]) -> None:
        try:
            import httpx
        except ImportError:
            raise ImportError(
                "httpx is required for ElasticAdapter: pip install httpx"
            )

        async def _do() -> None:
            lines: list[str] = []
            for item in items:
                lines.append(json.dumps({"index": {"_index": self.index}}))
                lines.append(json.dumps(_item_to_json(item)))
            body = "\n".join(lines) + "\n"
            headers: dict[str, str] = {"Content-Type": "application/x-ndjson"}
            if self.api_key:
                headers["Authorization"] = f"ApiKey {self.api_key}"
            async with httpx.AsyncClient() as client:
                resp = await client.post(
                    f"{self.url}/_bulk", content=body, headers=headers
                )
                if resp.status_code >= 400:
                    raise ExportError(
                        f"Elasticsearch export failed: {resp.status_code}"
                    )
                try:
                    payload = resp.json()
                except ValueError:
                    payload = None
                if isinstance(payload, dict) and payload.get("errors") is True:
                    raise ExportError(
                        f"Elasticsearch export failed: {resp.status_code}"
                    )

        await _with_retry(_do, self.retry)


def to_stix(
    alerts: list[Alert], ioc_matches: list[IocMatch] | None = None
) -> dict:
    """Convert alerts and optional IOC matches to a STIX 2.1 bundle."""
    objects: list[dict] = []
    for alert in alerts:
        objects.append(
            {
                "type": "indicator",
                "spec_version": "2.1",
                "id": f"indicator--{uuid.uuid4()}",
                "created": alert.triggered_at.isoformat(),
                "modified": alert.triggered_at.isoformat(),
                "name": alert.title,
                "description": alert.description,
                "pattern_type": "clawdstrike",
                "pattern": f"[alert:rule_name = '{_escape_stix_value(alert.rule_name)}']",
                "valid_from": alert.triggered_at.isoformat(),
                "labels": [alert.severity.value],
            }
        )
    if ioc_matches:
        for match in ioc_matches:
            for ioc in match.matched_iocs:
                objects.append(
                    {
                        "type": "indicator",
                        "spec_version": "2.1",
                        "id": f"indicator--{uuid.uuid4()}",
                        "created": match.event.timestamp.isoformat(),
                        "modified": match.event.timestamp.isoformat(),
                        "name": f"IOC: {ioc.indicator}",
                        "description": ioc.description
                        or f"IOC match: {ioc.indicator}",
                        "pattern_type": "stix",
                        "pattern": _ioc_to_stix_pattern(ioc),
                        "valid_from": match.event.timestamp.isoformat(),
                    }
                )
    return {
        "type": "bundle",
        "id": f"bundle--{uuid.uuid4()}",
        "objects": objects,
    }


def to_csv(items: list[Union[Alert, TimelineEvent]]) -> str:
    """Convert items to CSV string. Handles mixed Alert/TimelineEvent lists."""
    if not items:
        return ""
    alerts = [i for i in items if isinstance(i, Alert)]
    events = [i for i in items if isinstance(i, TimelineEvent)]
    sections: list[str] = []
    if alerts:
        sections.append(_alerts_csv(alerts))
    if events:
        sections.append(_events_csv(events))
    return "\n".join(sections)


def _alerts_csv(alerts: list[Alert]) -> str:
    headers = [
        "rule_name",
        "severity",
        "title",
        "triggered_at",
        "description",
        "evidence_count",
    ]
    rows = [
        ",".join(
            [
                _csv_escape(a.rule_name),
                _csv_escape(a.severity.value),
                _csv_escape(a.title),
                _csv_escape(a.triggered_at.isoformat()),
                _csv_escape(a.description),
                str(len(a.evidence)),
            ]
        )
        for a in alerts
    ]
    return "\n".join([",".join(headers)] + rows)


def _events_csv(events: list[TimelineEvent]) -> str:
    headers = [
        "timestamp",
        "source",
        "kind",
        "verdict",
        "summary",
        "process",
        "action_type",
    ]
    rows = [
        ",".join(
            [
                _csv_escape(e.timestamp.isoformat()),
                _csv_escape(e.source.value),
                _csv_escape(e.kind.value),
                _csv_escape(e.verdict.value),
                _csv_escape(e.summary),
                _csv_escape(e.process or ""),
                _csv_escape(e.action_type or ""),
            ]
        )
        for e in events
    ]
    return "\n".join([",".join(headers)] + rows)


def to_jsonl(items: list[Union[Alert, TimelineEvent]]) -> str:
    """Convert items to newline-delimited JSON (JSONL) string."""
    return "\n".join(json.dumps(_item_to_json(item)) for item in items)


def _item_to_json(item: Union[Alert, TimelineEvent]) -> dict:
    if isinstance(item, Alert):
        return {
            "type": "alert",
            "rule_name": item.rule_name,
            "severity": item.severity.value,
            "title": item.title,
            "triggered_at": item.triggered_at.isoformat(),
            "description": item.description,
            "evidence_count": len(item.evidence),
        }
    return {
        "type": "event",
        "timestamp": item.timestamp.isoformat(),
        "source": item.source.value,
        "kind": item.kind.value,
        "verdict": item.verdict.value,
        "summary": item.summary,
        "process": item.process,
        "action_type": item.action_type,
    }


def _escape_stix_value(value: str) -> str:
    """Escape backslashes, single quotes, and brackets for STIX patterns."""
    return value.replace("\\", "\\\\").replace("'", "\\'").replace("]", "\\]")


def _ioc_to_stix_pattern(ioc: IocEntry) -> str:
    ioc_type_val = ioc.ioc_type.value if hasattr(ioc.ioc_type, "value") else str(ioc.ioc_type)
    escaped = _escape_stix_value(ioc.indicator)
    mapping = {
        "sha256": f"[file:hashes.'SHA-256' = '{escaped}']",
        "sha1": f"[file:hashes.'SHA-1' = '{escaped}']",
        "md5": f"[file:hashes.MD5 = '{escaped}']",
        "domain": f"[domain-name:value = '{escaped}']",
        "ipv4": f"[ipv4-addr:value = '{escaped}']",
        "ipv6": f"[ipv6-addr:value = '{escaped}']",
        "url": f"[url:value = '{escaped}']",
    }
    return mapping.get(ioc_type_val, f"[x-clawdstrike:value = '{escaped}']")


def _csv_escape(value: str) -> str:
    if "," in value or '"' in value or "\n" in value or "\r" in value:
        return '"' + value.replace('"', '""') + '"'
    return value


__all__ = [
    "ExportAdapter",
    "RetryConfig",
    "WebhookAdapter",
    "SplunkHECAdapter",
    "ElasticAdapter",
    "to_stix",
    "to_csv",
    "to_jsonl",
]
